datafusion/common/src/dfschema.rs

```diff
@@ -116,14 +162,36 @@ impl DFSchema {
         )
     }
 
+    /// Assign identifier key groups
+    pub fn with_identifier_key_groups(
+        mut self,
+        identifier_key_groups: IdentifierKeyGroups,
+    ) -> Self {
+        self.identifier_key_groups = identifier_key_groups;
+        self
+    }
+
     /// Create a new schema that contains the fields from this schema followed by the fields
     /// from the supplied schema. An error will be returned if there are duplicate field names.
     pub fn join(&self, schema: &DFSchema) -> Result<Self> {
         let mut fields = self.fields.clone();
         let mut metadata = self.metadata.clone();
         fields.extend_from_slice(schema.fields().as_slice());
         metadata.extend(schema.metadata.clone());
-        Self::new_with_metadata(fields, metadata)
+        let mut id_key_groups = self.identifier_key_groups.clone();
```

Comment:
style nit is if you encapsulated this code in its own function (like Schema::join_functional_dependencies())` it might be more self-describing